### DISCUSSION TITLE
Autocompletion for undefined variables in expressions

### DISCUSSION THREAD

**Discussion by D8H** [4ian/GDevelop#3653] [Enhancement ideas (and their technical discussions)]
## Description

There is a preference to show in autocompletion the variables that are used in the event sheet but not defined (`useUndefinedVariablesInAutocompletion`). It was added in this PR: https://github.com/4ian/GDevelop/pull/2665
It's working only for variable fields but not for expression fields. Variables autocompletion in expressions was added in this PR: https://github.com/4ian/GDevelop/pull/2692

## Solution suggested

> 
> > I didn't find an easy way to use undefined variables without having them calculated at every input so I revert it.
> 
> Indeed, for undefined variables which require a more complex analysis (compared to "just" calling enumerateVariables), we might think about a cache/memoization of the results, with the challenge being (like all caches) to: 
> - properly invalidate the cache when needed 
> - and properly index the results with a proper "key". 
> 
> I think we could have this "cache" created at the initialization of a GenericExpressionField, and passed down to the autocompletions so that autocompletions of variables can be stored there.
> When a GenericExpressionField is blurred/destroyed, we release the cache.
> 
> This way we would compute variables "only" once per scene/project/object per field.
> 
> But all of this is if we want to support "undefined variables". Maybe we don't want to even do this in expressions?
> 
> When you write an expression, it might be acceptable to handle in autocompletions only the properly declared variables. It would encourage to declare them. And you always have the variables editor that will help you find undefined variables - and at this moment we can afford the "CPU power" to do it (because we only do it once per dialog, which is fair).

_Originally posted by @4ian in https://github.com/4ian/GDevelop/issues/2692#issuecomment-855241802_

## Alternatives considered
* Not doing it

**Comment by OverloadedTech** [comment]
👍
